libusb/os/windows_hotplug.c

```diff
@@ -0,0 +1,307 @@
+	struct winusb_device_priv *priv;
+
+	for_each_device_safe(ctx, dev, next_dev)
+	{
+		priv = usbi_get_device_priv(dev);
+		if(priv->hotplug_status != LEFT)
+		{
+			continue;
+		}
+
+		if(priv->initialized)
+		{
+			usbi_disconnect_device(dev);			
+		}
+		else
+		{
+			usbi_detach_device(dev);	
+		}
+	}
+
+	for_each_device(ctx, dev)
+	{
+		priv = usbi_get_device_priv(dev);
+
+		if(priv->hotplug_status != ARRIVED)
+		{
+			continue;
+		}
+
+		usbi_hotplug_notification(ctx, dev, LIBUSB_HOTPLUG_EVENT_DEVICE_ARRIVED);
+	}
+}
+
+static void windows_refresh_device_list_for_all_ctx(void)
+{
+	usbi_mutex_static_lock(&active_contexts_lock);
+
+	struct libusb_context *ctx;
+	for_each_context(ctx)
+	{
+		windows_refresh_device_list(ctx);
+	}
+	usbi_mutex_static_unlock(&active_contexts_lock);
+}
+
+#define WND_CLASS_NAME TEXT("libusb-1.0-windows-hotplug")
+
+static bool init_wnd_class(void)
+{
+	WNDCLASS wndClass = { 0 };
+	wndClass.lpfnWndProc = windows_proc_callback;
+	wndClass.hInstance = GetModuleHandle(NULL);
+	wndClass.lpszClassName = WND_CLASS_NAME;
+
+	if (!RegisterClass(&wndClass)) 
+	{
+		log_error("event thread: RegisterClass");
+		return false;
+	}
+
+	return true;
+}
+
+static DWORD WINAPI windows_event_thread_main(LPVOID lpParam)
+{
+	UNUSED(lpParam);
+
+	usbi_dbg(NULL, "windows event thread entering");
+
+	if (!init_wnd_class())
+	{
+		return (DWORD)-1;
+	}
+
+	windows_event_hWnd = CreateWindow(
+		WND_CLASS_NAME,
+		TEXT(""),
+		0, 
+		0, 0, 0, 0,
+		NULL, NULL,
+		GetModuleHandle(NULL),
+		NULL);
+
+	if (windows_event_hWnd == NULL)
+	{
+		log_error("event thread: CreateWindow");
+		return (DWORD)-1;
+	}
+
+	MSG msg;
+	BOOL ret_val;
+
+	while ((ret_val = GetMessage(&msg, windows_event_hWnd, 0, 0)) != 0)
+	{
+		if (ret_val == -1)
+		{
+			log_error("event thread: GetMessage");
+			break;
+		}
+
+		if(!SUCCEEDED(TranslateMessage(&msg)))
+		{
+			log_error("event thread: TranslateMessage");
+		}
+
+		if(!SUCCEEDED(DispatchMessage(&msg)))
+		{
+			log_error("event thread: DispatchMessage");
+		}
+	}
+
+	usbi_dbg(NULL, "windows event thread exiting");
+
+	return 0;
+}
+
+static bool register_device_interface_to_window_handle(
+	IN GUID interface_class_guid,
+	IN HWND hWnd,
+	OUT HDEVNOTIFY* device_notify_handle)
+{
+	DEV_BROADCAST_DEVICEINTERFACE notificationFilter = { 0 };
+	notificationFilter.dbcc_size = sizeof(notificationFilter);
+	notificationFilter.dbcc_devicetype = DBT_DEVTYP_DEVICEINTERFACE;
+	notificationFilter.dbcc_classguid = interface_class_guid; 
+
+	*device_notify_handle = RegisterDeviceNotification(
+		hWnd, 
+		&notificationFilter,
+		DEVICE_NOTIFY_WINDOW_HANDLE// | DEVICE_NOTIFY_ALL_INTERFACE_CLASSES// ?? for the ALL?
+	);
+
+	if (*device_notify_handle == NULL)
+	{
+		log_error("register_device_interface_to_window_handle");
+		return false;
+	}
+
+	return true;
+}
+
+static LRESULT CALLBACK windows_proc_callback(
+	HWND hWnd, 
+	UINT message, 
+	WPARAM wParam, 
+	LPARAM lParam)
+{
+	UNUSED(lParam);
+	
```

Comment:
Here and many other places there is trailing whitespace. If you run git show <hash> in a color terminal this will be highlighted in red. (Or git diff before committing.)

Reply:
Sorry, I don't get it precisely. Do you mean that there is a tab at line 260?